File: src/listeners/events/app_home_opened.py
from slack_sdk import WebClient
from logging import Logger
from database import get_user_prs
from helpers import get_status
from ..actions.handle_remove_pr import handle_remove_pr

# App home opened event callback
def app_home_opened_callback(client: WebClient, event: dict, logger: Logger):
    if event["tab"] != "home":
        return

    try:
        user_id = event["user"]
        logger.info(f"App home opened by user {user_id}.")
        
        # Fetch the user's active PRs
        active_prs = get_user_prs(user_id)

        blocks = [
            {
                "type": "section",
                "text": {
                    "type": "mrkdwn",
                    "text": f"*Welcome back, <@{user_id}>!* :house:"
                }
            },
            {
                "type": "divider"
            },
            {
                "type": "section",
                "text": {
                    "type": "mrkdwn",
                    "text": "*Your Active PRs*"
                }
            }
        ]

        if active_prs:
            logger.info(f"User {user_id} has {len(active_prs)} active PRs.")
            for pr in active_prs:
                pr_status = get_status(pr)
                reviewers_text = ""
                if pr['reviewers']:
                    reviewers_text = " - reviewed by "
                    reviewers_text += ", ".join(
                        [f"<@{r}>" for r in pr["reviewers"]]
                    )
                blocks.append({
                    "type": "section",
                    "text": {
                        "type": "mrkdwn",
                        "text": f"• *<{pr['permalink']}|{pr['name']}>* - {pr_status}{reviewers_text}"
                    }
                })
                blocks.append({
                    "type": "actions",
                    "elements": [
                        {
                            "type": "button",
                            "text": {
                                "type": "plain_text",
                                "text": "Remove"
                            },
                            "style": "danger",
                            "action_id": f"removePr_home_{pr['id']}",
                            "confirm": {
                                "title": {
                                    "type": "plain_text",
                                    "text": "Remove this PR?"
                                },
                                "text": {
                                    "type": "mrkdwn",
                                    "text": "This action cannot be undone."
                                },
                                "confirm": {
                                    "type": "plain_text",
                                    "text": "Yes, remove it"
                                },
                                "deny": {
                                    "type": "plain_text",
                                    "text": "Cancel"
                                }
                            }
                        }
                    ]
                })
        else:
            logger.info(f"No active PRs found for user {user_id}.")
            blocks.append({
                "type": "section",
                "text": {
                    "type": "mrkdwn",
                    "text": "You have no active PRs."
                }
            })

        # Publish the home view
        client.views_publish(
            user_id=user_id,
            view={
                "type": "home",
                "blocks": blocks
            }
        )
        logger.info(f"App home view published successfully for user {user_id}.")

    except Exception as e:
        logger.error(f"Error publishing home tab for user {user_id}: {e}")


# Handler for the remove button
def handle_remove_pr_home(ack, body, client, logger):
    ack()
    pr_id = body["actions"][0]["action_id"].split("_")[-1]  # Extract PR ID
    user_id = body["user"]["id"]
    
    try:
        handle_remove_pr(ack, body, client, logger)
        # Update the app home after removing the PR
        update_app_home(client, user_id, logger)

    except Exception as e:
        logger.error(f"Error removing PR {pr_id}: {e}")


# Function to republish the app home after changes
def update_app_home(client, user_id, logger):
    logger.info(f"Updating app home for user {user_id}")
    
    active_prs = get_user_prs(user_id)
    blocks = [
        {
            "type": "section",
            "text": {
                "type": "mrkdwn",
                "text": f"*Welcome back, <@{user_id}>!* :house:\n"
            }
        },
        {
            "type": "divider"
        },
        {
            "type": "section",
            "text": {
                "type": "mrkdwn",
                "text": "*Your Active PRs*"
            }
        }
    ]

    if active_prs:
        logger.info(f"User {user_id} has {len(active_prs)} active PRs.")
        for pr in active_prs:
            pr_status = get_status(pr)
            reviewers_text = ""
            if pr['reviewers']:
                reviewers_text = " - reviewed by "
                reviewers_text += ", ".join(
                    [f"<@{r}>" for r in pr["reviewers"]]
                )
            blocks.append({
                "type": "section",
                "text": {
                    "type": "mrkdwn",
                    "text": f"• *<{pr['permalink']}|{pr['name']}>* - {pr_status}{reviewers_text}"
                }
            })
            blocks.append({
                "type": "actions",
                "elements": [
                    {
                        "type": "button",
                        "text": {
                            "type": "plain_text",
                            "text": "Remove"
                        },
                        "style": "danger",
                        "action_id": f"removePr_home_{pr['id']}",
                        "confirm": {
                            "title": {
                                "type": "plain_text",
                                "text": "Remove this PR?"
                            },
                            "text": {
                                "type": "mrkdwn",
                                "text": "This action cannot be undone."
                            },
                            "confirm": {
                                "type": "plain_text",
                                "text": "Yes, remove it"
                            },
                            "deny": {
                                "type": "plain_text",
                                "text": "Cancel"
                            }
                        }
                    }
                ]
            })
    else:
        logger.info(f"No active PRs found for user {user_id}.")
        blocks.append({
            "type": "section",
            "text": {
                "type": "mrkdwn",
                "text": "You have no active PRs."
            }
        })

    client.views_publish(
        user_id=user_id,
        view={
            "type": "home",
            "blocks": blocks
        }
    )
    logger.info(f"App home view updated successfully for user {user_id}.")

File: src/listeners/events/AppHomeOpened.java
package prreminderbot.actions;

import com.slack.api.bolt.context.builtin.EventContext;
import com.slack.api.methods.SlackApiException;
import com.slack.api.Slack;
import com.slack.api.model.block.Blocks;
import com.slack.api.model.block.composition.BlockCompositions;
import com.slack.api.model.view.Views;
import org.slf4j.Logger;
import prreminderbot.database.Database;
import prreminderbot.helpers.Helpers;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;

public class AppHomeOpenedCallback {

    public static void handleAppHomeOpened(AppHomeOpenedEvent event, EventContext ctx, Slack client, Logger logger) {
        if (!event.getTab().equals("home")) {
            return;
        }

        try {
            String userId = event.getUser();
            logger.info("App home opened by user {}", userId);

            // Fetch the user's active PRs
            List<Map<String, Object>> activePrs = Database.getUserPrs(userId);
            List<com.slack.api.model.block.Block> blocks = new ArrayList<>();

            blocks.add(Blocks.section(section -> section.text(BlockCompositions.markdownText("*Welcome back, <@" + userId + ">!* :house:"))));
            blocks.add(Blocks.divider());
            blocks.add(Blocks.section(section -> section.text(BlockCompositions.markdownText("*Your Active PRs*"))));

            if (activePrs != null && !activePrs.isEmpty()) {
                logger.info("User {} has {} active PRs", userId, activePrs.size());
                for (Map<String, Object> pr : activePrs) {
                    String prStatus = Helpers.getStatus(pr);
                    String reviewersText = "";

                    if (pr.get("reviewers") != null) {
                        List<String> reviewers = (List<String>) pr.get("reviewers");
                        reviewersText = " - reviewed by " + String.join(", ", reviewers.stream().map(r -> "<@" + r + ">").toArray(String[]::new));
                    }

                    blocks.add(Blocks.section(section -> section.text(BlockCompositions.markdownText("• *<" + pr.get("permalink") + "|" + pr.get("name") + ">* - " + prStatus + reviewersText))));
                    blocks.add(Blocks.actions(actions -> actions.elements(List.of(
                        BlockElements.button(b -> b
                            .text(BlockCompositions.plainText("Remove"))
                            .style("danger")
                            .actionId("removePr_home_" + pr.get("id"))
                            .confirm(confirm -> confirm
                                .title(BlockCompositions.plainText("Remove this PR?"))
                                .text(BlockCompositions.markdownText("This action cannot be undone."))
                                .confirm(BlockCompositions.plainText("Yes, remove it"))
                                .deny(BlockCompositions.plainText("Cancel"))
                            )
                        )
                    ))));
                }
            } else {
                logger.info("No active PRs found for user {}", userId);
                blocks.add(Blocks.section(section -> section.text(BlockCompositions.markdownText("You have no active PRs."))));
            }

            // Publish the home view
            client.methods().viewsPublish(r -> r
                .userId(userId)
                .view(Views.view(view -> view.type("home").blocks(blocks)))
            );
            logger.info("App home view published successfully for user {}", userId);

        } catch (IOException | SlackApiException e) {
            logger.error("Error publishing home tab for user {}: {}", event.getUser(), e.getMessage());
        }
    }

    public static void handleRemovePrHome(ActionRequest req, EventContext ctx, Slack client, Logger logger) {
        try {
            String prId = req.getPayload().getActions().get(0).getActionId().split("_")[2];  // Extract PR ID
            String userId = req.getPayload().getUser().getId();
            handleRemovePr(req, ctx, client, logger);
            updateAppHome(client, userId, logger);
        } catch (IOException | SlackApiException e) {
            logger.error("Error removing PR: {}", e.getMessage());
        }
    }

    public static void updateAppHome(Slack client, String userId, Logger logger) throws IOException, SlackApiException {
        logger.info("Updating app home for user {}", userId);

        List<Map<String, Object>> activePrs = Database.getUserPrs(userId);
        List<com.slack.api.model.block.Block> blocks = new ArrayList<>();

        blocks.add(Blocks.section(section -> section.text(BlockCompositions.markdownText("*Welcome back, <@" + userId + ">!* :house:"))));
        blocks.add(Blocks.divider());
        blocks.add(Blocks.section(section -> section.text(BlockCompositions.markdownText("*Your Active PRs*"))));

        if (activePrs != null && !activePrs.isEmpty()) {
            logger.info("User {} has {} active PRs", userId, activePrs.size());
            for (Map<String, Object> pr : activePrs) {
                String prStatus = Helpers.getStatus(pr);
                String reviewersText = "";

                if (pr.get("reviewers") != null) {
                    List<String> reviewers = (List<String>) pr.get("reviewers");
                    reviewersText = " - reviewed by " + String.join(", ", reviewers.stream().map(r -> "<@" + r + ">").toArray(String[]::new));
                }

                blocks.add(Blocks.section(section -> section.text(BlockCompositions.markdownText("• *<" + pr.get("permalink") + "|" + pr.get("name") + ">* - " + prStatus + reviewersText))));
                blocks.add(Blocks.actions(actions -> actions.elements(List.of(
                    BlockElements.button(b -> b
                        .text(BlockCompositions.plainText("Remove"))
                        .style("danger")
                        .actionId("removePr_home_" + pr.get("id"))
                        .confirm(confirm -> confirm
                            .title(BlockCompositions.plainText("Remove this PR?"))
                            .text(BlockCompositions.markdownText("This action cannot be undone."))
                            .confirm(BlockCompositions.plainText("Yes, remove it"))
                            .deny(BlockCompositions.plainText("Cancel"))
                        )
                    )
                ))));
            }
        } else {
            logger.info("No active PRs found for user {}", userId);
            blocks.add(Blocks.section(section -> section.text(BlockCompositions.markdownText("You have no active PRs."))));
        }

        client.methods().viewsPublish(r -> r
            .userId(userId)
            .view(Views.view(view -> view.type("home").blocks(blocks)))
        );
        logger.info("App home view updated successfully for user {}", userId);
    }
}
